### PR TITLE
Support message property filters + new operators

### PR DESCRIPTION
**Pull request by prolic**
see https://github.com/prooph/event-store/pull/286
see https://github.com/prooph/event-store/issues/284

### CODE REVIEW

**Comment by codeliner** [comment]
@oqq and @basz do you have time to take a second look?

LGTM but maybe you see some flaws that I don't see.

File: src/MySqlEventStore.php [hunk 1]
```diff
@@ -569,18 +575,56 @@ private function createWhereClauseForMetadata(?MetadataMatcher $metadataMatcher)
...
+            } elseif (Operator::NOT_IN()->is($operator)) {
+                $operatorString = 'NOT IN (';
+                $operatorStringEnd = ')';
+            } else {
+                $operatorString = $operator->getValue();
             }
 
-            $where[] = "metadata->\"$.$field\" $operator $parameter";
-            $values[$parameter] = $value;
+            if (FieldType::METADATA()->is($fieldType)) {
```
Comment (basz) [inline src/MySqlEventStore.php]: Why not; 

Operator->is(...) for eadability?

File: src/MySqlEventStore.php [hunk 2]
```diff
@@ -569,18 +575,56 @@ private function createWhereClauseForMetadata(?MetadataMatcher $metadataMatcher)
...
         foreach ($metadataMatcher->data() as $key => $match) {
+            $fieldType = $match['fieldType'];
             $field = $match['field'];
-            $operator = $match['operator']->getValue();
+            $operator = $match['operator'];
             $value = $match['value'];
-            $parameter = ':metadata_'.$key;
+            $parameters = [];
+
+            if (is_array($value)) {
```
Comment (basz) [inline src/MySqlEventStore.php]: Make array if not array, then loop through?

File: src/PostgresEventStore.php [hunk 3]
```diff
@@ -551,22 +557,59 @@ private function createWhereClauseForMetadata(?MetadataMatcher $metadataMatcher)
...
             }
 
-            if (is_int($value)) {
-                $where[] = "CAST(metadata->>'$field' as int) $operator $parameter";
-                $values[$parameter] = $value;
+            $parameterString = implode(', ', $parameters);
+
+            $operatorStringEnd = '';
+
+            if (Operator::REGEX()->is($operator)) {
```
Comment (basz) [inline src/PostgresEventStore.php]: Operator->is(...)

File: src/MySqlEventStore.php (hunk 1 above)
Comment (prolic) [inline src/MySqlEventStore.php]: It's field type, not operator :)

File: src/MySqlEventStore.php (hunk 2 above)
Comment (prolic) [inline src/MySqlEventStore.php]: ok

File: src/PostgresEventStore.php (hunk 3 above)
Comment (prolic) [inline src/PostgresEventStore.php]: ok

File: src/MySqlEventStore.php (hunk 2 above)
Comment (prolic) [inline src/MySqlEventStore.php]: ah not ok, it's different behaviour, see the code below

File: src/PostgresEventStore.php (hunk 3 above)
Comment (prolic) [inline src/PostgresEventStore.php]: problem is auto complete missing then

Comment (basz) [inline src/PostgresEventStore.php]: `/** @var FQCN $operator */`? you against that?

**Pull request merged**